components/places/src/api/places_api.rs

```diff
@@ -0,0 +1,298 @@
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc, Mutex,
+};
+use sync15::{telemetry, ClientInfo};
+
+#[repr(u8)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum ConnectionType {
+    ReadOnly = 1,
+    ReadWrite = 2,
+    Sync = 3,
+}
+
+impl ConnectionType {
+    pub fn from_primitive(p: u8) -> Option<Self> {
+        match p {
+            1 => Some(ConnectionType::ReadOnly),
+            2 => Some(ConnectionType::ReadWrite),
+            3 => Some(ConnectionType::Sync),
+            _ => None,
+        }
+    }
+}
+
+impl ConnectionType {
+    pub fn rusqlite_flags(&self) -> OpenFlags {
+        let common_flags = OpenFlags::SQLITE_OPEN_NO_MUTEX | OpenFlags::SQLITE_OPEN_URI;
+        match self {
+            ConnectionType::ReadOnly => common_flags | OpenFlags::SQLITE_OPEN_READ_ONLY,
+            ConnectionType::ReadWrite => {
+                common_flags | OpenFlags::SQLITE_OPEN_CREATE | OpenFlags::SQLITE_OPEN_READ_WRITE
+            }
+            ConnectionType::Sync => common_flags | OpenFlags::SQLITE_OPEN_READ_WRITE,
+        }
+    }
+}
+
+// We only allow a single PlacesApi per filename.
+// XXX - probably need encryption key here too so we can do something sane
+// if we attempt to open the same file with different keys.
+lazy_static! {
+    static ref APIS: Mutex<HashMap<PathBuf, Arc<PlacesApi>>> = Mutex::new(HashMap::new());
+}
+
+static ID_COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+struct SyncState {
+    conn: PlacesDb,
+    client_info: Cell<Option<ClientInfo>>,
+}
+
+/// The entry-point to the places API. This object gives access to database
+/// connections and other helpers. It enforces that only 1 write connection
+/// can exist to the database at once.
+pub struct PlacesApi {
+    db_name: PathBuf,
+    encryption_key: Option<String>,
+    // XXX - this should probably be a Cell instead of a mutex as we will
+    // rely on external synchronization.
+    write_connection: Mutex<Option<PlacesDb>>,
+
+    sync_state: Mutex<Option<SyncState>>,
+    id: usize,
+}
+
+impl PlacesApi {
+    /// Create a new, or fetch an already open, PlacesApi backed by a file on disk.
+    pub fn new(db_name: impl Into<PathBuf>, encryption_key: Option<&str>) -> Result<Arc<Self>> {
+        let db_name = db_name.into().canonicalize()?;
+        Self::new_or_existing(db_name, encryption_key)
+    }
+
+    /// Create a new, or fetch an already open, memory-based PlacesApi. You must
+    /// provide a name, but you are still able to have a single writer and many
+    ///  reader connections to the same memory DB open.
+    pub fn new_memory(db_name: &str, encryption_key: Option<&str>) -> Result<Arc<Self>> {
+        let name = PathBuf::from(format!("file:{}?mode=memory&cache=shared", db_name));
+        Self::new_or_existing(name, encryption_key)
+    }
+
+    fn new_or_existing(db_name: PathBuf, encryption_key: Option<&str>) -> Result<Arc<Self>> {
+        // XXX - we should check encryption_key via the HashMap here too.
```

Comment:
nit: Your previous comments about not wanting the keys in memory forever here make sense - so maybe update this comment to reflect that we don't want to it and that this might not behave as expected if the encryption keys differ?

Reply:
Makes more sense to not store the encryption key anymore given that we don't use it for anything right now.

In the future we could compare with a hash of the key or something if we want.

Reply:
I think what we'd want is for opening the connection to require the key as well, or something.